crates/github-actions-expressions/src/lib.rs

```diff
@@ -43,6 +43,156 @@ impl PartialEq<str> for Function<'_> {
     }
 }
 
+/// Represents a function call in a GitHub Actions expression.
+#[derive(Debug, PartialEq)]
+pub struct Call<'src> {
+    /// The function name, e.g. `foo` in `foo()`.
+    pub func: Function<'src>,
+    /// The function's arguments.
+    pub args: Vec<SpannedExpr<'src>>,
+}
+
+impl<'src> Call<'src> {
+    /// Performs constant evaluation of a GitHub Actions expression
+    /// function call.
+    fn consteval(&self) -> Option<Evaluation> {
+        let args = self
+            .args
+            .iter()
+            .map(|arg| arg.consteval())
+            .collect::<Option<Vec<Evaluation>>>()?;
+
+        match &self.func {
+            f if f == "format" => Self::consteval_format(&args),
+            f if f == "contains" => Self::consteval_contains(&args),
+            f if f == "startsWith" => Self::consteval_startswith(&args),
+            f if f == "endsWith" => Self::consteval_endswith(&args),
+            f if f == "toJSON" => Self::consteval_tojson(&args),
+            f if f == "fromJSON" => Self::consteval_fromjson(&args),
+            _ => None,
+        }
+    }
+
+    /// Constant-evaluates a `format(fmtspec, args...)` call.
+    ///
+    /// See: <https://github.com/actions/languageservices/blob/1f3436c3cacc0f99d5d79e7120a5a9270cf13a72/expressions/src/funcs/format.ts>
+    fn consteval_format(args: &[Evaluation]) -> Option<Evaluation> {
+        if args.is_empty() {
+            return None;
+        }
+
+        let format_str = &args[0];
+        let format_template = format_str.to_string();
+
+        let mut result = format_template;
+
+        // Replace {0}, {1}, {2}, etc. with the corresponding arguments
+        for (i, arg) in args.iter().skip(1).enumerate() {
+            let placeholder = format!("{{{}}}", i);
+            result = result.replace(&placeholder, &arg.to_string());
+        }
+
+        Some(Evaluation::String(result))
+    }
```

Comment:
@mostafa FYI, I think this and some of the other consteval function implementations don't 100% match the reference implementation (which I've linked in the comments for each function). 

Some of these will be nontrivial to implement, so we can make a judgement call here 🙂 -- I'd personally be okay with having some of the more annoying ones return `None` for now so that we can get this out as an MVP, and then iterate on their implementations going forwards. But I'm curious what you think.

(The consequence of that will be that we sometimes won't have a full evaluation to offer for a constant expression, so we won't have a fix in those cases. I think that's okay for the MVP.)

Reply:
How about matching the reference implementation as much as we can, while we are at it? 🤔

Reply:
Works for me!